src/box/lua/tuple.c

```diff
@@ -318,7 +326,8 @@ lbox_tuple_new(lua_State *L)
 	 * box.tuple.new(1, 2, 3).
 	 */
 	box_tuple_format_t *fmt = box_tuple_format_default();
-	if (argc != 1 || (!lua_istable(L, 1) && !luaT_istuple(L, 1))) {
+	if (box_tuple_new_arg_list &&
+	    (argc != 1 || (!lua_istable(L, 1) && !luaT_istuple(L, 1)))) {
```

Comment:
> box.tuple.new(value)
>
> Construct a new tuple from either a scalar or a Lua table.

https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_tuple/new/

Passing one scalar **is** documented behavior. It shouldn't be changed.

Reply:
BTW, this if-condition is already quite big and now it becomes even bigger. How about moving it into a variable?

```c
bool is_scalar = !lua_istable(L, 1) && !luaT_istuple(L, 1); 
bool args_as_fields = argc == 1 ? is_scalar :
        !box_tuple_new_forbid_vararg;
if (args_as_fields) {
        <...>
```

Reply:
Actually, allowing a sclar really complicates the code. Your fix won't work, because the branch below creates a tuple using a default format.

Reply:
I don't think we need to support a scalar argument. A tuple is an array of scalars so it should be initialized with an array - I don't see why we should handle the case when there's just one element in a special way. I'd just update the documentation instead. After all, this commit is already marked as a breaking change.

Reply:
> I see no motivation for the second change.

1. A tuple is essentially a vector (i.e., non-scalar) object.
2. Supporting scalar arguments prior to this patch was seamless, now it will complicate the code a lot (we will need another branch for handling scalars), so we pay maintenance here.
3. Why would one hesitate writing `box.tuple.new{a}` over `box.tuple.new(a)`?

Reply:
2. We just a introduced a new backward compatibility option, so what’s the problem?

Reply:
I don't think that we can't do it. I just don't see a motivation for it:

1. The old behavior doesn't forbid introducing a new feature with a good API.
2. The old behavior isn't beautiful, but a beauty should never be counted as the motivation for incompatibilities. (Unless there is a major redesign plan, which will make users cry with happiness.)
3. There are words about maintenante cost, but the question is about 30-40 lines of code that is concentrated in one place. Doesn't sound as a problem. At worst -- just write it twice and it will not be a big problem. It is not something that goes over, say, five layers of abstractions.

> so what’s the problem?

That's the wrong question. We should ask ourself 'what is the problem to keep the updating process smooth?', not 'is there a formal rule that would stop me from making updates harder?'. If you want a rule: don't do unmovitated breaking changes even in major releases. Even under options, flags and so on.

There is a big difference between 'dnf upgrade --> profit' and 'dnf upgrade -> google error message -> look around the code -> check the documentation -> ask in the support chat -> got somehow that it can be solved by an option -> tweak the option -> got another error message ...'. The first scenario involves just devops team. The second engages the application development team, which is possibly in another organization.

If you imagine that this is a few minutes problem, that's not. A month if you're lucky. (If you're interecting, ask our management how easy the transitions from 2.8/2.10 to 2.11 are going on.)

Imagine that after those adventures the developers from those organizations will come to us and ask: what is the wonderful feature that costs all these stones? Your answer?

Reply:
Discusses this again with product manager. It seems that (and I agree) that we should break only what we really need. We need options in the second argument so we have to break tuple creation by two and more scalar arguments. But we should accept scalar in the first field. Perhaps it's not very useful but still user-friendly. And compatible with `space:select`.
So I propose to allow to create tuples with `new(1, {opt=2})`. If we want to continue discussion we should summon product manager too.

Reply:
@Totktonada and @alyapunov thanks a lot for your feedback! I learned and understood a lot from this discussion, I would even create a Wiki page about making breaking changes based on what @Totktonada wrote.

Fixed, added support for scalar argument.